scripts/create-functionmap.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+$filter = function (string $name): bool {
+    $namespaces = ['MongoDB\BSON\\', 'MongoDB\Driver\\'];
+
+    foreach ($namespaces as $namespace) {
+        if (str_starts_with($name, $namespace) || str_starts_with($name, strtolower($namespace))) {
+            return true;
+        }
+    }
+
+    return false;
+};
+
+$getFunctionCallMapEntry = function(ReflectionFunctionAbstract $function): array {
+    $returnType = match(true) {
+        $function->hasReturnType() => (string) $function->getReturnType(),
+        $function->hasTentativeReturnType() => (string) $function->getTentativeReturnType(),
+        default => 'void',
+    };
+    $callmapEntry = [$returnType];
+
+    foreach ($function->getParameters() as $parameter) {
+        $parameterKey = $parameter->getName();
+        if ($parameter->isOptional()) {
+            $parameterKey .= '=';
+        }
+
+        $parameterType = (string) $parameter->getType();
+        if ($function->getName() === 'unserialize' && $parameter->getName() === 'serialized') {
+            $parameterType = 'string';
+        }
```

Comment:
I took a trip through the PHP source code to figure out what's going on.

In PHP < 8.0, the parameter had no type and was called `$serialized`. In PHP 8.0, the parameter name was changed to `$data`, and a string type added. Since the PHP 7 interface has no type, we can't add our own type in the method signature directly, which is why there's a `@param` annotation. This in turn causes our definition to be without a parameter type, which is why an empty type is returned.

There are two options for this: we can add a conditional method signature for PHP >= 8, which fixes the parameter name and adds the `string` type. As these maps are generated on PHP 8 only, this would ensure correct signature generation without a workaround. The other option would be to add a comment explaining this and doing away with it once we drop the `Serializable` interface. `Serializable` is only used on PHP < 7.4, but one could argue that dropping the interface in a minor release could be a BC break. With all of this in mind, adding conditional method signatures seems like a sensible choice and doesn't cause any follow-up work outside of dropping the interface whenever we're ready to do so.

Reply:
> The other option would be to add a comment explaining this and doing away with it once we drop the Serializable interface. Serializable is only used on PHP < 7.4, but one could argue that dropping the interface in a minor release could be a BC break

I agree that it'd be a BC break to remove Serializable. I'm already tracking that in [PHPC-2248] with a 2.0 fixVersion.

> With all of this in mind, adding conditional method signatures seems like a sensible choice and doesn't cause any follow-up work outside of dropping the interface whenever we're ready to do so.

Switching to a conditional declaration sounds good, especially if that would simplify this script. I think that deserves its own PHPC ticket for visibility. If you want to merge this as-is you can just make a note to come back and clean up this workaround later.

Reply:
Updated to contain reworked method signatures for `Serializable::unserialize`.